feat: configure tracing layers

therustmonk opened the pull request:
## Description

This PR adds the ability to configure (add) **custom tracing layers**, such as OpenTelemetry. For now, this extended functionality is available in the Optimism CLI.

Implementation details:

1. The `Layers` type is public and can be modified using the `.add_layer()` method.
2. The `Tracer` trait can now be configured with a custom set of layers via a special method `init_with_layers()`. The original `init()` method is preserved but internally reuses the new one by providing an empty `Layers` list, just like before.
3. A new `configure()` method has been added to the Optimism `Cli`, which provides a builder-wrapper `CliApp`. This allows users to specify their own `CliRunner` or `Layers`. It also stores the tracing guard within the same structure.
4. The Optimism `Cli` now fully relies on `CliRunner`.
5. To avoid duplicating a large number of `launcher` type definitions, a special `Launcher` trait has been introduced. This allows defining a custom launcher using a struct, adding flexibility:
   * All contextual data for launching can be stored in the struct and reused.
   * A launcher can now be constructed even from a `json::Value` (e.g., read from a file).
   * Launchers can be layered, wrapping one another and forwarding the launch method.
   * Different methods can be added for various stages of the launcher's lifecycle.
   * For compatibility, a closure-style launcher wrapper `FnLauncher` is provided. It is also used in the previous `Cli` methods of the Optimism node.   
6. A new `reth-tracing-otlp` crate has been added to provide an OpenTelemetry tracing layer for exporting spans in OTLP format.

## Example

1. Now the parameters allow you to obtain a `CliApp` (via the `configure()` method, which provides access to tracing layers).

```rust
let mut cli_app = Cli::<OpChainSpecParser, OpRbuilderArgs>::parse().configure();
```

2. The OTLP layer is available in a separate crate, `reth_tracing_otlp`. The layer is created using the `layer()` function, which expects a service name as input.

```diff
+ let otlp = reth_tracing_otlp::layer("my-op-node");
+ cli_app.access_tracing_layers()?.add_layer(otlp);
```

3. The only thing is, I added a `Launcher` trait to avoid duplicating the long type parameter definitions from `Cli` in the `CliApp` function. Now a wrapper launcher like `FnLauncher` is required, but in the long run, this paves the way to make `Launcher` configurable and decomposable.

```rust
cli_app.run(FnLauncher::new::<OpChainSpecParser, OpRbuilderArgs>(
    |builder, builder_args| async move {
        let _span = tracing::span!(Level::INFO, "Node");
        tracing::info!("Starting the node");
        let rollup_args = builder_args.rollup_args;
        let op_node = OpNode::new(rollup_args.clone());
        let handle = builder
            .with_types::<OpNode>()
            .with_components(op_node.components())
            .with_add_ons(OpAddOnsBuilder::default().build())
            // .on_node_started(move |_ctx| Ok(()))
            .launch()
            .await?;
        handle.node_exit_future.await
    },
))?;
```

mattsse commented on crates/cli/commands/src/launcher.rs:
```diff
@@ -0,0 +1,52 @@
+use reth_node_builder::{NodeBuilder, WithLaunchContext};
+use std::{fmt, marker::PhantomData, sync::Arc};
+
+#[async_trait(?Send)]
```
do we even need async_trait here anymore? 
could maybe even use `fn entrypoint<'a> -> impl Future`?

therustmonk replied:
Implemented using `impl Future`,`async_trait` dependency removed.

mattsse reviewed (commented):
cool, I like this a lot already, supportive

therustmonk commented:
The PR is ready for review. The opentelemetry integration is implemented in a separate crate in order to avoid breaking `reth-tracing` while also making this feature configurable in the future.

<img width="2047" alt="Screenshot 2025-05-16 at 17 09 35" src="https://github.com/user-attachments/assets/0b412e80-db45-4ff7-9ff7-8385144f7b0b" />

Also it's now possible to add a custom implementation of `TracingLayer` to analyze spans.

----

Jaeger started with:

```sh
docker run --rm --name jaeger \
  -p 16686:16686 \
  -p 4317:4317 \
  -p 4318:4318 \
  -p 5778:5778 \
  -p 9411:9411 \
  jaegertracing/jaeger:2.5.0
```

mattsse commented on crates/cli/commands/src/launcher.rs:
```diff
@@ -0,0 +1,53 @@
+use reth_node_builder::{NodeBuilder, WithLaunchContext};
+use std::{fmt, marker::PhantomData, sync::Arc};
+
+pub trait Launcher<C, Ext>
+where
+    C: ChainSpecParser,
+    Ext: clap::Args + fmt::Debug,
+{
+    fn entrypoint<'a>(
+        self,
+        builder: WithLaunchContext<NodeBuilder<Arc<DatabaseEnv>, C::ChainSpec>>,
+        builder_args: Ext,
+    ) -> impl Future<Output = eyre::Result<()>>
+    where
+        Ext: 'a;
+}
+
+pub struct FnLauncher<F, Fut> {
```
can we add a oneliner doc for trait,entrypoint and FnLauncher

mattsse reviewed (changes requested):
ty!

this makes a ton of sense,

this looks great I only have some smol doc nitpicks

I like the CliApp type a lot and I have a feeling that this pattern could be useful to simplify some things in the builder stack as well

fyi @onbjerg

therustmonk commented:
@mattsse Thanks for the review. I added comments and also removed `'a` lifetime from the `entrypoint()` method (it was added for `async_trait` before).

mattsse commented on crates/cli/commands/src/node.rs:
```diff
@@ -138,11 +139,10 @@ where
     where
-        L: FnOnce(WithLaunchContext<NodeBuilder<Arc<DatabaseEnv>, C::ChainSpec>>, Ext) -> Fut,
-        Fut: Future<Output = eyre::Result<()>>,
+        L: Launcher<C, Ext>,
```
this is so much better

mattsse reviewed (approved):
lgtm!
